### PR TITLE
Allow PrivateUse1 backends to opt out of CEA decomposition

### LINKED ISSUE #188052: Allow PrivateUse1 backends to opt out of CompositeExplicitAutograd decomposition
## Problem

PyTorch's dispatch table resolves **CompositeExplicitAutograd (CEA)** kernels before the backend fallback. PrivateUse1 backends that register a single catch-all fallback kernel never see composite ops — the dispatcher silently decomposes them into primitives before the fallback is invoked.

The dispatch resolution order for a PrivateUse1 backend is:

```
1. PrivateUse1 kernel          ← direct registration, always wins
2. CEANF alias kernel          ← silently decomposes (CompositeExplicitAutogradNonFunctional)
3. CEA alias kernel            ← silently decomposes (CompositeExplicitAutograd)  ← THE PROBLEM
4. CIA / Autograd alias        ← CompositeImplicitAutograd
5. Backend fallback            ← only reached if no alias kernel matched above
```

As a result, ops like the following are decomposed before the backend ever sees them:

| Composite op      | Decomposes to              | Alias key |
| ----------------- | -------------------------- | --------- |
| `convolution`     | `convolution_overrideable` | CEA       |
| `softmax.int_out` | `softmax_out`              | CEA       |

Steps 2 and 3 cannot currently be bypassed without registering a PrivateUse1-specific kernel for every affected op individually. Step 4 (CIA) is a separate mechanism and is intentionally not part of this proposal.

## Impact

**Who is affected:** PrivateUse1 backends that implement a single catch-all fallback kernel and forward ATen ops verbatim to an external execution environment — for example, a remote device, a hardware accelerator with its own op set, or a tracing backend. These backends need to receive composite ops whole; the external side often has optimized implementations for ops like `convolution` directly, not just the decomposed primitives like `convolution_overrideable`.

**Example pattern:**

```python
# A PrivateUse1 backend registers one fallback to handle all ops.
lib = torch.library.Library("_", "IMPL", "PrivateUse1")

def my_fallback(op, args, kwargs):
    # forward op to external execution environment (remote device, accelerator, etc.)
    external_device.run(op.name(), args, kwargs)

lib.fallback(my_fallback)
```

With this pattern, calling `torch.nn.functional.conv2d(tensor)` on a PrivateUse1 tensor results in `my_fallback` receiving `convolution_overrideable` — not `convolution`. If the external execution environment expects `convolution`, the backend breaks silently.

**Scale:**

Hundreds of ATen op overloads have CEA registrations. A proxy backend must manually register a stub for each one to prevent decomposition — a list that grows silently with every PyTorch release.

## Request

A backend-level opt-out mechanism so that PrivateUse1 backends can bypass CEA/CEANF decompositions and receive composite ops directly through the registered fallback. 

I am working on a fix and will raise a PR shortly.

cc: @mansiag05 

### Alternatives

_No response_

### Additional context

_No response_

cc @NmomoN @mengpenghui @fwenguang @cdzhan @1274085042 @PHLens @albanD

### PR DESCRIPTION
**Pull request by vishalgoyal316**
## Background

Fixes #188052

PyTorch's dispatch table resolves CEA/CEANF kernels before the backend fallback. PrivateUse1 backends that register a single catch-all fallback never see composite ops intact — e.g. `convolution` is silently decomposed to `convolution_overrideable` (a stub that raises `NotImplementedError`) before the fallback fires. There is no way to bypass this without registering a per-op stub for every CEA-aliased op, a list that grows silently with each PyTorch release.

## Solution

A new atomic flag `privateuse1_skip_cea_` on the `Dispatcher`, checked during `computeDispatchTableEntryWithDebug()`. When set, alias lookups for steps 2.1 (CEANF) and 2.2 (CEA) are skipped for PrivateUse1, letting the dispatch table entry fall through to the backend fallback. The flag is written once under the dispatcher mutex along with a full table refresh — **zero hot-path overhead**.

Returns a RAII handle whose destructor restores the previous behavior and refreshes the table. `CompositeImplicitAutograd` (CIA) is intentionally unaffected — skipping it would break autograd for ops like `conv2d` and `layer_norm`.

```python
handle = torch.utils.skip_cea_decomposition_for_privateuse1()
torch._C._dispatch_privateuse1_skip_cea_enabled()  # True
del handle  # restores previous behavior
```

## Testing

Two new test classes in `test/test_dispatch.py`:

- **`TestPrivateUse1SkipCEAFlag`** — flag/RAII semantics: enable, query, RAII restore, double-registration guard, re-enable after release.
- **`TestPrivateUse1SkipCEATableInspection`** — dispatch table: CEA slot changes on skip, CIA slot unaffected, table restored after release, CPU/CUDA slots unaffected.

All 11 new tests pass. No physical device required.

### CODE REVIEW

**Comment by vishalgoyal316** [comment]
cc: @mansiag05
